erigon-lib/commitment/hex_patricia_hashed.go

```diff
@@ -950,10 +1102,231 @@ func (hph *HexPatriciaHashed) needUnfolding(hashedKey []byte) int {
+					if !bytes.Equal(subTrieRoot, cellHash[1:]) {
+						return nil, fmt.Errorf("subTrieRoot(%x) != cellHash(%x)", subTrieRoot, cellHash[1:])
+					}
+					// // DEBUG patch with cell hash which we know to be correct
+					// nextNode = trie.NewHashNode(cellHash[1:])
+				}
+			}
+		} else if cellToExpand.storageAddrLen > 0 { // storage cell
+			storageUpdate, err := hph.Ctx.Storage(cellToExpand.storageAddr[:cellToExpand.storageAddrLen])
+			if err != nil {
+				return nil, err
+			}
+			storageValueNode := trie.ValueNode(storageUpdate.Storage[:storageUpdate.StorageLen])
+			nextNode = &storageValueNode //nolint:ineffassign, wastedassign
+			break
+		} else if cellToExpand.accountAddrLen > 0 { // account cell
+			accNode, err := hph.createAccountNode(cellToExpand, row, hashedKey, codeReads)
+			if err != nil {
+				return nil, err
+			}
+			nextNode = accNode
+			keyPos++ // only move one nibble
+		} else if cellToExpand.hashLen > 0 { // hash cell means we will expand using a full node
+			nextNode = &trie.FullNode{}
+			keyPos++
+		} else if cellToExpand.IsEmpty() {
+			nextNode = nil // no more expanding can happen (this could be due )
+		} else { // default for now before we handle extLen
+			nextNode = &trie.FullNode{}
+			keyPos++
+		}
+
+		// process the current node
+		if fullNode, ok := currentNode.(*trie.FullNode); ok { // handle full node case
+			for col := 0; col < 16; col++ {
+				currentCell := &hph.grid[row][col]
+				if currentCell.IsEmpty() {
+					fullNode.Children[col] = nil
+					continue
+				}
+				cellHash, _, _, err := hph.computeCellHashWithStorage(currentCell, hph.depths[row], nil)
+				if err != nil {
+					return nil, err
+				}
+				fullNode.Children[col] = trie.NewHashNode(cellHash[1:]) // because cellHash has 33 bytes and we want 32
+			}
+			fullNode.Children[currentNibble] = nextNode // ready to expand next nibble in the path
+		} else if accNode, ok := currentNode.(*trie.AccountNode); ok {
+			if len(hashedKey) <= 64 { // no storage, stop here
+				nextNode = nil // nolint:ineffassign, wastedassign
+				break
+			}
+			// there is storage so we need to expand further
+			accNode.Storage = nextNode
+		} else if extNode, ok := currentNode.(*trie.ShortNode); ok { // handle extension node case
+			// expect only one item in this row, so take the first one
+			// technically it should be at the last nibble of the key but we will adjust this later
+			if extNode.Val != nil { // early termination
+				break
+			}
+			extNode.Val = nextNode
+		} else {
+			break // break if currentNode is nil
+		}
+		// we need to check if we are dealing with the next node being an account node and we have a storage key,
+		// in that case start a new tree for the storage
+		if nextAccNode, ok := nextNode.(*trie.AccountNode); ok && len(hashedKey) > 64 {
+			nextNode = &trie.FullNode{}
+			nextAccNode.Storage = nextNode
+		}
+		currentNode = nextNode
+	}
+	tr := trie.NewInMemoryTrie(rootNode)
+	return tr, nil
```

Comment:
what if in next iteration we will encode everything as-we-see into witness format instead of temporal intermediate representation?

Reply:
I don't think you can do that without at least some form of auxiliary memory structures to guide you in the building process, and also another layer doing the encoding process. Emitting the serialized witness operands directly sounds super inconvenient, and would also be a nightmare to debug if something goes wrong.

So I think the intermediate representation is useful, and what potentially could be accomplished in future iterations is to build it inside the implementations of `fold` and `unfold`.

Reply:
yeah i meant those as a follow up later on